components/flash_memory/flash_nvs.c

```diff
@@ -1,98 +1,90 @@
 // Copyright 2022 PWr in Space, Krzysztof Gliwiński
+// Edited Mateusz Kłosiński 21.02.2024
 #include "flash_nvs.h"
 
 #define TAG "NVS"
 
 static struct {
-  nvs_handle_t handle;
-  bool initialized;
+    nvs_handle_t handle;
+    bool initialized;
 } nvs;
 
-NVSResult NVS_init(void) {
-  esp_err_t res;
-
-  res = nvs_flash_init();
-  if (res != ESP_OK) {
-    ESP_LOGE(TAG, "NVS error init %s", esp_err_to_name(res));
-    return NVS_INIT_ERROR;
-  }
-
-  nvs.handle = 0;
-  return NVS_OK;
-}
-
-NVSResult NVS_write_uint8(const char* key, uint8_t val) {
-  esp_err_t res;
-  res = nvs_open("storage", NVS_READWRITE, &nvs.handle);
-  if (res != ESP_OK) {
-    ESP_LOGE(TAG, "NVS open error %s", esp_err_to_name(res));
-    return NVS_OPEN_ERROR;
-  }
-
-  res = nvs_set_u8(nvs.handle, key, val);
-  if (res != ESP_OK) {
+NVSResult NVS_init(NVSData* data_array, size_t length) {
+    esp_err_t res;
+
+    res = nvs_flash_init();
+    if (res != ESP_OK) {
+        ESP_LOGE(TAG, "NVS error init %s", esp_err_to_name(res));
+        return NVS_INIT_ERROR;
+    }
+
+    if (length == 0 || data_array == NULL) {
+        length = 0;
+        ESP_LOGE(TAG, "data array is length 0 %s", esp_err_to_name(res));
+        return NVS_INIT_ERROR;
+    }
+
+    int32_t tmp_Val;
+    for (int i = 0; i < length; i++) {
+        NVSResult result = NVS_read_int32t(data_array[i].key, &tmp_Val);
+        if (result == NVS_OK && tmp_Val == data_array[i].value) {
+            ESP_LOGI(TAG, "Key %s already exists in NVS, skipping", data_array[i].key);
+        } else if (result == NVS_OK && tmp_Val != data_array[i].value) {
+            ESP_LOGI(TAG, "Key %s exists in NVS but we are saving different value, saving",
+                     data_array[i].key);
+            NVS_write_int32(data_array[i].key, data_array[i].value);
+        } else if (result == NVS_READ_ERROR) {
+            ESP_LOGI(TAG, "Key %s doesnt exist in NVS, saving", data_array[i].key);
```

Comment:
Korzystasz z błedu, który zwraca funkcja że nie udało się odczytać. Ale to nie znaczy że nie ma klucza, jest kilka innych przyczyn przez które może coś takiego się pojawić.
W funkcji NVS_read_int32, przy nvs_set_i32 sprawdzasz czy zwórciło coś innego niż ESP_OK, a z [dokumentacji](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/api-reference/storage/nvs_flash.html#_CPPv410nvs_get_i812nvs_handle_tPKcP6int8_t) widać, żę nie tylko błędny klucz zwraca coś innego niż ESP_OK. Więc aby uniknąć kłopotów w szczególnych przypadkach. Trzeba reagować tylko na błąd związany z brakiem klucza.